components/Buttons/src/MDCFloatingButton.m

```diff
@@ -60,79 +75,369 @@ - (instancetype)initWithFrame:(CGRect)frame shape:(MDCFloatingButtonShape)shape
+    }
+    if ([aDecoder containsValueForKey:MDCFloatingButtonContentEdgeInsetsDictionaryKey]) {
+      _shapeToContentEdgeInsets
+          = [aDecoder decodeObjectForKey:MDCFloatingButtonContentEdgeInsetsDictionaryKey];
+    }
+    if ([aDecoder containsValueForKey:MDCFloatingButtonHitAreaInsetsDictionaryKey]) {
+      _shapeToHitAreaInsets
+          = [aDecoder decodeObjectForKey:MDCFloatingButtonHitAreaInsetsDictionaryKey];
+    }
   }
   return self;
 }
 
 - (void)encodeWithCoder:(NSCoder *)aCoder {
   [super encodeWithCoder:aCoder];
-  [aCoder encodeInteger:_shape forKey:MDCFloatingButtonShapeKey];
+  [aCoder encodeInteger:self.shape forKey:MDCFloatingButtonShapeKey];
+  [aCoder encodeObject:self.shapeToMinimumSize forKey:MDCFloatingButtonMinimumSizeDictionaryKey];
+  [aCoder encodeObject:self.shapeToMaximumSize forKey:MDCFloatingButtonMaximumSizeDictionaryKey];
+  [aCoder encodeObject:self.shapeToContentEdgeInsets
+                forKey:MDCFloatingButtonContentEdgeInsetsDictionaryKey];
+  [aCoder encodeObject:self.shapeToHitAreaInsets
+                forKey:MDCFloatingButtonHitAreaInsetsDictionaryKey];
 }
 
 #pragma mark - UIView
 
-- (CGSize)sizeThatFits:(__unused CGSize)size {
+- (CGSize)intrinsicContentSize {
+  NSLog(@"intrinsicContentSize");
   switch (_shape) {
     case MDCFloatingButtonShapeDefault:
       return CGSizeMake([[self class] defaultDimension], [[self class] defaultDimension]);
     case MDCFloatingButtonShapeMini:
       return CGSizeMake([[self class] miniDimension], [[self class] miniDimension]);
     case MDCFloatingButtonShapeLargeIcon:
       return CGSizeMake([[self class] defaultDimension], [[self class] defaultDimension]);
+    case MDCFloatingButtonShapeExtendedTrailingIcon:
+    case MDCFloatingButtonShapeExtendedLeadingIcon: {
+      const CGSize intrinsicTitleSize
+          = [self.titleLabel sizeThatFits:CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX)];
+      const CGSize intrinsicImageSize
+          = [self.imageView sizeThatFits:CGSizeMake(CGFLOAT_MAX, CGFLOAT_MAX)];
+      CGFloat intrinsicWidth = intrinsicTitleSize.width + intrinsicImageSize.width
+          + self.imageTitlePadding + self.contentEdgeInsets.left + self.contentEdgeInsets.right;
+      CGFloat intrinsicHeight = MAX(intrinsicTitleSize.height, intrinsicImageSize.height)
+          + self.contentEdgeInsets.top + self.contentEdgeInsets.bottom;
+      return CGSizeMake(intrinsicWidth, intrinsicHeight);
+    }
   }
 }
 
-- (CGSize)intrinsicContentSize {
-  switch (_shape) {
-    case MDCFloatingButtonShapeDefault:
-      return CGSizeMake([[self class] defaultDimension], [[self class] defaultDimension]);
-    case MDCFloatingButtonShapeMini:
-      return CGSizeMake([[self class] miniDimension], [[self class] miniDimension]);
-    case MDCFloatingButtonShapeLargeIcon:
-      return CGSizeMake([[self class] defaultDimension], [[self class] defaultDimension]);
+//- (CGSize)sizeThatFits:(CGSize)size {
```

Comment:
This is an alternate solution that would just add the `imageTitlePadding` value to the size returned by super. However, this has risks by relying on UIButton (or MDCButton) to perform sizing in a compatible manner.